### ISSUE TITLE
Improving TadGAN performance on personal dataset

### ISSUE THREAD

**Issue by cjtaylo-csu** [sintel-dev/Orion#295]
* Orion version: 0.2.1.dev0
* Python version: 3.6.9
* Operating System: Ubuntu 18.04.6 LTS

### Description

When applying TadGAN to my personal data-set, I am seeing very low performance.
I have followed the following issues and blog posts to apply TadGAN/orion appropriately to my data-set.
https://github.com/sintel-dev/Orion/issues/231
https://github.com/sintel-dev/Orion/issues/162
https://github.com/sintel-dev/Orion/issues/160
https://github.com/sintel-dev/Orion/issues/227
https://github.com/sintel-dev/Orion/issues/224

Besides changing the "interval" hyper-parameter, and preprocessing my data, are there any other recommendations?

I believe perhaps more training is required, but for 35 epochs of training on my 4M+ data-point training set, the training took 13+ hours on my NVIDIA GTX1080 GPU.

### What I Did

I am using basically the stock [tadgan_gpu pipeline](https://github.com/sintel-dev/Orion/blob/master/orion/pipelines/verified/tadgan/tadgan_gpu.json), but with the time-series aggregate.interval changed to match my dataset.

#### Pre-Processing

Below is a screenshot of my data after preprocessing. 

timestamp | value
-- | --
2089 | 0.2468387653
2104 | 0.2471801744
2119 | 0.2475215835
2134 | 0.2478629925
2149 | 0.2482044016
2164 | 0.2485458107
2180 | 0.2488872198
2195 | 0.2490579243
2210 | 0.2492286289
2225 | 0.2495700379

You mention in https://github.com/sintel-dev/Orion/issues/227:
"timestamp column needs to be of integer type, preferably a unix timestamp." 
Do I need to reformat my data to a unix timestamp? I am currently using the timestamp above with an interval hyper-parameter of "45".

#### Results

I confirmed that when training, my GPU was in use by using the `nvidia-smi` tool.

I also tried training on the 1st 560,000 data-points of my data-set, as this length is more similar to the Nasa-MSL data-set used in the paper, but the performance was just as poor.

Attached is an image showing the missed anomalies in my test dataset, and sole anomaly detection which only partially overlaps the true anomaly.
![image](https://user-images.githubusercontent.com/95731513/169867566-d4cd10e5-becf-4919-99ee-1a94cdb1bcf0.png)

I dont believe TadGAN is appropriately training on my dataset given the loss plot throughout training in the image below.
![image](https://user-images.githubusercontent.com/95731513/169867678-05c170a2-890f-4112-9252-7b8b6dcef2db.png)

Thank you for your help!

**Comment by sarahmish** [comment]
Hi @cjtaylo-csu, this is awesome!

I'll try to cover as much as I can from the questions you asked, if I forget to touch on something please let me know!

### adjusting preprocessing to speedup runtime
4M+ datapoints is a lot of information. You can choose to reduce the sample size by increasing the interval, so for example if you double it you would end up with ~2M data points where you aggregate them by the mean value as default (you can change this to max, min, etc).

### UnixTimestamp v.s. Integers
It is true that the expectation for the timestamp column to be of unix timestamp value, however it should not differ if you use regular integer values are actual timestamps.

### detection performance
the testing dataset seems to have many anomalies, therefore you want to make the `find_anomalies` primitive very sensitive to the discrepancies between the original and reconstructed signal. I recommend playing around with that primitive. In terms of training TadGAN itself, you can plot the reconstructed signal and compare it with the original to visually assess the training performance. You can use [tadgan_viz](https://github.com/sintel-dev/Orion/blob/master/orion/pipelines/verified/tadgan/tadgan_viz.json) pipeline for this task then use `anomalies, viz = orion.detect(signal, visualization=True)` where `viz` is now a dictionary with `generated_timeseries` key. 

I hope I answered all your questions! I know TadGAN takes more time to train because it contains 4 networks. I recommend trying other pipelines as well!

**Comment by cjtaylo-csu** [comment]
Sarah,

Thank you very much for your detailed response. I will look into the find_anomalies primitive and the tadgan_viz pipeline.

Thanks again

Chris

**Comment by cjtaylo-csu** [comment]
Hi Sarah,

I am trying to experiment with tadgan_viz pipeline.
Are there any examples using this pipeline?

I got an error when directly applying the tadgan_viz pipeline to the NYC_Taxi data (In the Tulog example):
`ValueError: Error when checking input: expected input_33 to have 3 dimensions, but got array with shape (64, 1)`

Instead I appended the following from the tadgan_viz pipeline to the tulog_gpu pipeline, and was able to train on the data.
```
    "outputs": {
        "default": [
            {
                "name": "events",
                "variable": "orion.primitives.timeseries_anomalies.find_anomalies#1.y"
            }
        ],
        "visualization": [
            {
                "name": "generated_timeseries",
                "variable": "orion.primitives.tadgan.score_anomalies#1.predictions"
            }
        ]
    }
```

But I don't quite understand the returned viz dictionary.
For the NYC_TAXI data which tadgan was trained upon, the returned dictionary (viz) from the line:
`anomalies, viz = orion.detect(df, visualization=True)`

appears to be a 2D array of 860 x 5 elements. I don't understand as NYC_TAXT (df) is of length 1032.

Thank you for the clarification
 
Chris
